# Library import

In [37]:
# Standard library imports (general utilities)
import os            # File and path operations
from typing import Optional, Dict, List, Tuple  # Type hints for readability

# Third‑party libraries (data science stack)
import numpy as np   # Numerical computing
import pandas as pd  # Data analysis and tabular processing
import matplotlib.pyplot as plt  # Plotting (will use for EDA visuals)

# Display options for clarity during EDA
pd.set_option("display.max_columns", 120)   # Show more columns in tables
pd.set_option("display.width", 160)         # Wider console width for tables

#### File path definition

In [38]:
# Base folder where raw data is stored
DATA_DIR = "./data/raw"

# Dictionary with key = dataset name, value = path to CSV
FILES = {"train_operational_readouts_csv": os.path.join(DATA_DIR, "train_operational_readouts.csv"), 
         "validation_operational_readouts": os.path.join(DATA_DIR, "validation_operational_readouts.csv"), 
         "test_operational_readouts": os.path.join(DATA_DIR, "test_operational_readouts.csv"), 
         "train_specifications": os.path.join(DATA_DIR, "train_specifications.csv"), 
         "validation_specifications": os.path.join(DATA_DIR, "validation_specifications.csv"), 
         "test_specifications": os.path.join(DATA_DIR, "test_specifications.csv"), 
         "train_tte": os.path.join(DATA_DIR, "train_tte.csv"), 
         "validation_labels": os.path.join(DATA_DIR, "validation_labels.csv"), 
         "test_labels": os.path.join(DATA_DIR, "test_labels.csv")}

for name, path in FILES.items(): # Quick check: print all paths
    print(f"{name:35} -> {path}")

train_operational_readouts_csv      -> ./data/raw\train_operational_readouts.csv
validation_operational_readouts     -> ./data/raw\validation_operational_readouts.csv
test_operational_readouts           -> ./data/raw\test_operational_readouts.csv
train_specifications                -> ./data/raw\train_specifications.csv
validation_specifications           -> ./data/raw\validation_specifications.csv
test_specifications                 -> ./data/raw\test_specifications.csv
train_tte                           -> ./data/raw\train_tte.csv
validation_labels                   -> ./data/raw\validation_labels.csv
test_labels                         -> ./data/raw\test_labels.csv


### Save csv files loader

In [39]:
def safe_read_csv(path: str, nrows: int | None = None) -> pd.DataFrame | None:
#  Safely reads a CSV. Returns None if the file does not exist or cannot be read.
    if not os.path.exists(path):
        print(f"[safe_read_csv] File does not exist: {path}")
        return None
    
    try:
        return pd.read_csv(path, low_memory=False, nrows=nrows)
    except Exception as exc:
        print(f"[safe_read_csv] Error reading {path}: {exc}")
        return None


#### Train_operational_readouts stats and EDA

In [40]:
DATA_DIR = "../data/raw" # Path to local file
train_ops_path = os.path.join(DATA_DIR, "train_operational_readouts.csv")
train_ops_df = safe_read_csv(train_ops_path, nrows=50_000) # Load a manageable sample (50k rows)

In [41]:
train_ops_df.shape

(50000, 107)

In [54]:
train_ops_df.head(206)

,vehicle_id,time_step,171_0,666_0,427_0,837_0,167_0,167_1,167_2,167_3,167_4,167_5,167_6,167_7,167_8,167_9,309_0,272_0,272_1,272_2,272_3,272_4,272_5,272_6,272_7,272_8,272_9,835_0,370_0,291_0,291_1,291_2,291_3,291_4,291_5,291_6,291_7,291_8,291_9,291_10,158_0,158_1,158_2,158_3,158_4,158_5,158_6,158_7,158_8,158_9,100_0,459_0,459_1,459_2,459_3,459_4,459_5,459_6,459_7,459_8,459_9,459_10,459_11,459_12,459_13,459_14,459_15,459_16,459_17,459_18,459_19,397_0,397_1,397_2,397_3,397_4,397_5,397_6,397_7,397_8,397_9,397_10,397_11,397_12,397_13,397_14,397_15,397_16,397_17,397_18,397_19,397_20,397_21,397_22,397_23,397_24,397_25,397_26,397_27,397_28,397_29,397_30,397_31,397_32,397_33,397_34,397_35
0,0,11.2,167985.0,10787.0,7413813.0,2296.0,4110.0,1296420.0,1628265.0,630345.0,1269525.0,4772940.0,2706706.0,222225.0,6240.0,0.0,70.0,1435083.0,857662.0,384579.0,668642.0,7239843.0,398490.0,3887.0,0.0,0.0,0.0,8036751.0,0.0,1227.0,555.0,463.0,925.0,468.0,225.0,535.0,516.0,492.0,729.0,66.0,97056.0,2690052.0,2945268.0,788437.0,687480.0,595164.0,491232.0,532932.0,809628.0,505693.0,858410.0,203.676778,111.911500,147.265389,200.479944,230.306278,277.722417,315.748806,372.164528,864.246250,920.881111,637.901639,744.618944,880.866889,1272.323972,1847.623667,940.785694,2.900083,0.208444,0.056417,0.058444,446956.0,411420.0,203024.0,26636.0,29156.0,7616.0,449537.0,233352.0,139920.0,12648.0,2813.0,224.0,53161.0,178881.0,138250.0,13328.0,3581.0,88.0,16361.0,131601.0,116541.0,13506.0,2856.0,48.0,6337.0,105412.0,95728.0,15609.0,1984.0,8.0,784.0,150228.0,261904.0,93172.0,17874.0,452.0
1,0,11.4,167985.0,10787.0,7413813.0,2296.0,4111.0,1302855.0,1628265.0,630345.0,1269526.0,4772940.0,2706706.0,222225.0,6240.0,0.0,70.0,1440661.0,857662.0,384579.0,668642.0,7239843.0,398490.0,3887.0,0.0,0.0,0.0,8040811.0,0.0,1230.0,558.0,463.0,925.0,469.0,226.0,535.0,516.0,493.0,729.0,66.0,97056.0,2693100.0,2947368.0,788437.0,687480.0,595164.0,491232.0,532932.0,809628.0,505693.0,860571.0,204.256750,112.924250,147.265389,201.479944,230.306278,277.722417,315.748806,372.164528,864.246250,920.881111,637.901639,745.618944,880.866889,1272.323972,1847.623667,940.785694,2.900083,0.208444,0.056417,0.058444,446964.0,411420.0,203027.0,26638.0,29157.0,7616.0,451193.0,233354.0,139920.0,12649.0,2813.0,224.0,53210.0,178883.0,138252.0,13328.0,3582.0,88.0,16368.0,131601.0,116542.0,13507.0,2856.0,48.0,6339.0,105413.0,95729.0,15610.0,1984.0,8.0,784.0,150228.0,261905.0,93172.0,17874.0,452.0
2,0,19.6,331635.0,14525.0,13683604.0,2600.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,70.0,1787736.0,1133132.0,598351.0,1167062.0,12314224.0,460240.0,3887.0,0.0,0.0,0.0,12777022.0,0.0,2136.0,954.0,850.0,1420.0,722.0,412.0,880.0,666.0,586.0,1143.0,162.0,181632.0,4249020.0,4630440.0,1539133.0,1421172.0,1039764.0,749472.0,740724.0,995796.0,574045.0,1379191.0,321.671972,157.312500,193.792833,263.577611,310.711861,366.149250,415.642472,484.391167,1146.111611,1286.536333,900.062917,1123.232556,1449.545611,2140.037472,5046.748278,1151.010139,3.320194,0.218806,0.056778,0.058444,756665.0,647348.0,286811.0,30967.0,31213.0,7745.0,633790.0,423395.0,271940.0,16190.0,3573.0,232.0,75038.0,352791.0,327992.0,17325.0,4451.0,92.0,24028.0,234737.0,216619.0,17000.0,3476.0,48.0,12055.0,167693.0,142900.0,19263.0,2441.0,12.0,1420.0,204832.0,313485.0,106464.0,19306.0,452.0
3,0,20.2,354975.0,15015.0,14540449.0,2616.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,70.0,1824409.0,1166074.0,634595.0,1233908.0,13275730.0,466753.0,3887.0,0.0,0.0,0.0,13612083.0,0.0,2218.0,1014.0,892.0,1471.0,749.0,425.0,901.0,702.0,589.0,1197.0,174.0,193728.0,4462548.0,4988028.0,1696022.0,1565484.0,1112544.0,789228.0,774588.0,1015104.0,576901.0,1428606.0,331.479028,162.731639,198.104472,269.712889,320.087333,377.478667,425.901361,495.749583,1173.882583,1323.460972,923.099361,1161.893139,1501.973944,2208.782833,5587.856667,1160.593833,3.336417,0.218806,0.056778,0.058444,812577.0,686860.0,302955.0,31927.0,31488.0,7749.0,651902.0,478279.0,292109.0,16755.0,3753.0,232.0,77118.0,39408

In [ ]:
train_tte_path = os.path.join(DATA_DIR, "train_tte.csv")
train_tte_df = safe_read_csv(train_tte_path) 


In [48]:
train_tte_df.head(100)

,vehicle_id,length_of_study_time_step,in_study_repair
0,0,510.0,0
1,2,281.8,0
2,3,293.4,0
3,4,210.0,0
4,5,360.4,0
...,...,...,...
95,132,208.4,0
96,134,351.8,1
97,135,490.0,0
98,136,303.4,0


In [ ]:
train_ops_df.dtypes.value_counts()

float64    106
int64        1
Name: count, dtype: int64

> 106 float64 values are operational values (masked sensors or histograms)

In [ ]:

nan_percentage = (train_ops_df.isna().mean() * 100).sort_values(ascending=False)  # Missing values (top 10 columns with most NA)
print("\nMissing values (%):")
print(nan_percentage.head(100))


Missing values (%):
167_2    1.224
167_3    1.224
167_0    1.224
167_4    1.224
167_1    1.224
         ...  
272_8    0.032
272_0    0.032
272_1    0.032
272_4    0.032
272_5    0.032
Length: 100, dtype: float64


In [ ]:
if "vehicle_id" in train_ops_df.columns and "time_step" in train_ops_df.columns:
    print("\nUnique vehicles:", train_ops_df["vehicle_id"].nunique())
    print("Min time_step:", train_ops_df["time_step"].min())
    print("Max time_step:", train_ops_df["time_step"].max())


Unique vehicles: 668
Min time_step: 0.0
Max time_step: 507.4


 - Each row = vehicle snapshot.
  - 2 types of features:
        Accumulative counters (increases value according to time_step).
        Bins Histograms (capture a variable distribution).
- NaN values, do not imply error, represent no data registered on that bin.
→ Can be speccificly replaced with a imputation plan (can be 0).

- Different scales (0.05 vs milions).
→ Its neccesary to use scalers or proportional converters to avoid modeling errors.